### PR TITLE
Add checking for corresponding pages for link tags

### LINKED ISSUE #6710: AMP - Only show canonical links to different page versions when present
Once https://github.com/zeit/next.js/pull/6706 is merged, we need to update these sections of code to only include these elements when the other version exists:

https://github.com/zeit/next.js/blob/ff5c8465deef35eb087237bbf912dab739337bcd/packages/next/pages/_document.js#L204

https://github.com/zeit/next.js/blob/ff5c8465deef35eb087237bbf912dab739337bcd/packages/next/pages/_document.js#L243

### PR DESCRIPTION
**Pull request by ijjk**
This adds a check for correspondings pages e.g. `dog.js` and `dog.amp.js` so we can create the correct `rel="canonical"` and only add `rel="amphtml"` when an AMP page is present. 

Closes: #6710

### CODE REVIEW

File: packages/next/pages/_document.js [hunk 1]
```diff
@@ -201,7 +203,7 @@ export class Head extends Component {
               name="viewport"
               content="width=device-width,minimum-scale=1,initial-scale=1"
             />
-            <link rel="canonical" href={page} />
+            {<link rel="canonical" href={page.split('.amp')[0] + (hasCanonical ? '' : '?amp=1')} />}
```
Comment (Timer) [inline packages/next/pages/_document.js]: This should just be -- this element can be excluded all together when there's no canonical version AFAIK:
```suggestion
            {hasCanonical && <link rel="canonical" href={page} />}
```

**Review by Timer — CHANGES_REQUESTED** [review]

**Review by Timer — DISMISSED** [review]
If tests pass and no significant changes are necessary to get them passing.

File: packages/next-server/server/require.ts [hunk 2]
```diff
@@ -51,7 +51,18 @@ export function getPagePath(page: string, distDir: string, opts: PagePathOptions
...
+  let hasAmpCanonical = false
+  try {
+    hasAmpCanonical = Boolean(
+      getPagePath(page, distDir, { amphtml: !isAmp }),
+    )
+  } catch (_) {}
+
+  return {
+    mod: require(pagePath),
+    [isAmp ? 'hasCanonical' : 'hasAmp']: hasAmpCanonical,
```
Comment (dav-is) [inline packages/next-server/server/require.ts]: I think `hasComplement` would make it more clear than `hasAmpCanonical`

File: packages/next/pages/_document.js [hunk 3]
```diff
@@ -201,7 +203,7 @@ export class Head extends Component {
               name="viewport"
               content="width=device-width,minimum-scale=1,initial-scale=1"
             />
-            <link rel="canonical" href={page} />
+            <link rel="canonical" href={page.split('.amp')[0]} />
```
Comment (dav-is) [inline packages/next/pages/_document.js]: What's this result in? Seems weird

File: packages/next/pages/_document.js [hunk 4]
```diff
@@ -150,6 +150,8 @@ export class Head extends Component {
       ampEnabled,
       styles,
       amphtml,
+      hasAmp,
+      // hasCanonical,
```
Comment (dav-is) [inline packages/next/pages/_document.js]: Why's this commented out?

File: packages/next/server/on-demand-entry-handler.js [hunk 5]
```diff
@@ -234,7 +234,11 @@ export default function onDemandEntryHandler (devMiddleware, multiCompiler, {
       const absolutePagePath = pagePath.startsWith('next/dist/pages') ? require.resolve(pagePath) : join(pagesDir, pagePath)
 
       page = posix.normalize(pageUrl)
-      const result = { isAmp, pathname: page }
+      const result = {
+        isAmp,
+        pathname: page,
+        [isAmp ? 'hasCanonical' : 'hasAmp']: Boolean(await findPageFile(pagesDir, normalizedPagePath, pageExtensions, !isAmp, ampEnabled))
```
Comment (dav-is) [inline packages/next/server/on-demand-entry-handler.js]: Does this still need the try {} catch {}?

File: packages/next/pages/_document.js (hunk 3 above)
Comment (ijjk) [reply packages/next/pages/_document.js]: It takes `/dog.amp` and turns it into `/dog`

Comment (dav-is) [reply packages/next/pages/_document.js]: `/dog/lab.amp` => ?

Comment (dav-is) [reply packages/next/pages/_document.js]: Hmm I guess it just seems strange to have a relative href in a canoncial tag

Comment (Timer) [reply packages/next/pages/_document.js]: `/dog/lab.amp` will be => `/dog/lab`

File: packages/next/server/on-demand-entry-handler.js (hunk 5 above)
Comment (Timer) [reply packages/next/server/on-demand-entry-handler.js]: `findPageFile` doesn't throw

**Review by Timer — APPROVED** [review]
We can keep revising these behaviors in follow up PRs, but let's :shipit: for now.

File: packages/next/pages/_document.js (hunk 3 above)
Comment (dav-is) [reply packages/next/pages/_document.js]: Didn't realize `page` was the path

**Review by dav-is — APPROVED** [review]

**Comment by egdelwonk** [comment]
with the usage of withAmp would you need to account for an associated amp page differently ? cc @Timer

**Pull request merged**